src/util.h

```diff
@@ -178,77 +178,75 @@ inline TypeName* Unwrap(v8::Local<v8::Object> object);
 
 inline void SwapBytes(uint16_t* dst, const uint16_t* src, size_t buflen);
 
-class Utf8Value {
+// Allocates an array of member type T. For up to StackStorageSize items,
+// the stack is used, otherwise malloc().
+template<typename T, size_t StackStorageSize>
+class MaybeStackBuffer {
   public:
-    explicit Utf8Value(v8::Isolate* isolate, v8::Local<v8::Value> value);
-
-    ~Utf8Value() {
-      if (str_ != str_st_)
-        free(str_);
+    T* operator*() {
+      return buf_;
     }
 
-    char* operator*() {
-      return str_;
-    };
-
-    const char* operator*() const {
-      return str_;
-    };
+    const T* operator*() const {
+      return buf_;
+    }
 
     size_t length() const {
       return length_;
-    };
-
-  private:
-    size_t length_;
-    char* str_;
-    char str_st_[1024];
-};
+    }
 
-class TwoByteValue {
-  public:
-    explicit TwoByteValue(v8::Isolate* isolate, v8::Local<v8::Value> value);
+    // Call to make sure enough space for `size` entries is available.
+    // There can only be 1 call to AssureSufficientStorage per instance.
+    void AssureSufficientStorage(size_t size) {
+      if (size <= StackStorageSize) {
+        buf_ = buf_st_;
+      } else {
+        buf_ = static_cast<T*>(malloc(sizeof(T) * size));
+        CHECK_NE(buf_, nullptr);
+      }
+    }
 
-    ~TwoByteValue() {
-      if (str_ != str_st_)
-        free(str_);
+    void SetLength(size_t length) {
+      length_ = length;
     }
 
-    uint16_t* operator*() {
-      return str_;
-    };
+    void SetLengthAndZeroTerminate(size_t length) {
+      SetLength(length);
 
-    const uint16_t* operator*() const {
-      return str_;
-    };
+      // T() is 0 for integer types, nullptr for pointers, etc.
+      buf_[length] = T();
+    }
 
-    size_t length() const {
-      return length_;
-    };
+    void AllocateAndZeroTerminate(size_t length) {
+      AssureSufficientStorage(length + 1);
+      SetLengthAndZeroTerminate(length);
+    }
+
+    MaybeStackBuffer() : length_(0), buf_(nullptr) {}
```

Comment:
Can I suggest you do `buf_(buf_st_)` and set `buf_[0] = T()`?  That would let you get rid of all the `AllocateAndZeroTerminate(0)` calls.